cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringClientHandler.java

```diff
@@ -0,0 +1,120 @@
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import javax.inject.Inject;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link io.cdap.http.HttpHandler} to manage tethering client v3 REST APIs
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class TetheringClientHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TetheringClientHandler.class);
+  private static final Gson GSON = new Gson();
+  static final String CREATE_TETHER = "/v3/tethering/connect";
+
+  private final TetheringStore store;
+  private final CConfiguration cConf;
+  private final String instanceName;
+
+  @Inject
+  TetheringClientHandler(CConfiguration cConf, TetheringStore store) {
+    this.store = store;
+    this.cConf = cConf;
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tethering.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      try {
+        RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+      } catch (Exception e) {
+        LOG.error("Failed to set default authenticator", e);
+      }
+    }
+  }
+
+  /**
+   * Initiates tethering with the server.
+   */
+  @POST
+  @Path("/tethering/create")
+  public void createTethering(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetheringCreationRequest tetheringCreationRequest = GSON.fromJson(content, TetheringCreationRequest.class);
+
+    PeerInfo peer = null;
+    try {
+      peer = store.getPeer(tetheringCreationRequest.getPeer());
+    } catch (PeerNotFoundException e) {
+      // Do nothing, expected if peer is not already configured.
+    }
+    if (peer != null) {
+      LOG.info("Peer {} is already present in state {}, ignoring tethering request",
+               peer.getName(), peer.getTetheringStatus());
+      responder.sendStatus(HttpResponseStatus.OK);
+      return;
+    }
+
+    List<NamespaceAllocation> namespaces = tetheringCreationRequest.getNamespaceAllocations();
+    TetheringConnectionRequest tetheringConnectionRequest = new TetheringConnectionRequest(instanceName,
+                                                                                           namespaces);
+    if (tetheringCreationRequest.getEndpoint() == null) {
+      throw new BadRequestException("Endpoint is null");
+    }
+    URI endpoint = new URI(tetheringCreationRequest.getEndpoint());
+    HttpResponse response = TetheringUtils.sendHttpRequest(HttpMethod.POST, endpoint.resolve(CREATE_TETHER),
```

Comment:
I feel like this request should be done in the agent service so that it can be retried correctly. It is better to just the peer information to the local store, and let the agent to contact the server side to keep the states in-sync.

Reply:
done